bee-common/bee-common/src/event.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright 2020 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! A module that provides a generic, type-safe event bus for arbitrary event types.
+
+use dashmap::DashMap;
+use std::any::{Any, TypeId};
+
+type Listener<'a> = dyn Fn(&dyn Any) + Send + Sync + 'a;
+
+/// An event bus for arbitrary event types.
+pub struct Bus<'a, ID = TypeId> {
+    listeners: DashMap<TypeId, Vec<(Box<Listener<'a>>, ID)>>,
+}
+
+impl<'a, ID> Default for Bus<'a, ID> {
+    fn default() -> Self {
+        Self {
+            listeners: DashMap::default(),
+        }
+    }
+}
+
+impl<'a, ID: Clone + PartialEq> Bus<'a, ID> {
+    /// Dispatch an event via is this event bus.
+    ///
+    /// All active listeners registered for this event will be invoked.
+    pub fn dispatch<E: Any>(&self, event: E) {
+        if let Some(mut ls) = self.listeners.get_mut(&TypeId::of::<E>()) {
+            ls.iter_mut().for_each(|(l, _)| l(&event))
+        }
+    }
+
+    /// Add an event listener bound to a specific event type, `E`, and registered
```

Comment:
I think the 120 columns policy should apply to comments as well.

Reply:
Will change them.